com.unity.multiplayer.mlapi/Runtime/Core/SnapshotRTT.cs

```diff
@@ -0,0 +1,94 @@
+using System;
+using UnityEngine;
+
+namespace MLAPI
+{
+    internal class ClientRtt
+    {
+        /// <summary>
+        /// Round-trip-time data
+        /// </summary>
+        public struct Rtt
+        {
+            public double Best;
+            public double Average;
+            public double Worst;
+            public int SampleCount;
+        }
+        public ClientRtt()
+        {
+            m_RttSendTimes = new double[k_RingSize];
+            m_SendKey = new int[k_RingSize];
+            m_MeasuredLatencies = new double[k_RingSize];
+        }
+
+        /// <summary>
+        /// Returns the Round-trip-time computation for this client
+        /// </summary>
+        public Rtt GetRtt()
+        {
+            var ret = new Rtt(); // is this a memory alloc ? How do I get a stack alloc ?
+            var index = m_LatenciesBegin;
+            double total = 0.0;
+            ret.Best = m_MeasuredLatencies[m_LatenciesBegin];
+            ret.Worst = m_MeasuredLatencies[m_LatenciesBegin];
+
+            while (index != m_LatenciesEnd)
```

Comment:
I'll bet there is a Google interview question that asks how to do this operation in constant time :)

Reply:
I'm familiar with keeping a running sum and subtracting the element that are removed at the other end of the ring buffer. Then, you can divide by the number of elements.

However, I don't know of a good way to do it for the `best` and `worst` entries. I believe we can only do `O(log(n))` but that would require an associative container for a handful of latencies. Seemed Overkill. Is it possible your comment forgot to consider min/max latency?

Reply:
For min/max just keep track of the min/max and then compare < & > when adding / removing from the ring buffer?

Reply:
So, let's say stored latencies are `[10, 60, 20, 60, 11, 80 ]`, min is 10, max is 80, a new latency of 15 needs to be stored. We now have `[60, 20, 60, 11, 80, 15]` (not in those positions, though). I can remove the `min: 10`, I can keep the `max 80`, but finding `min: 11` requires a traversal.

Reply:
I guess you could use 2 priority queues or heaps, one for min.  But to your point we can save this for benchmarking